toolchain/sem_ir/type_iterator.h

```diff
@@ -0,0 +1,174 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEM_IR_TYPE_ITERATOR_H_
+#define CARBON_TOOLCHAIN_SEM_IR_TYPE_ITERATOR_H_
+
+#include <concepts>
+#include <utility>
+#include <variant>
+
+#include "llvm/ADT/SmallVector.h"
+#include "toolchain/sem_ir/file.h"
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::SemIR {
+
+// Iterates over the the types in a given instruction. For the purposes of the
+// iterator, a facet value and an interface can also be treated like a type and
+// be iterated over.
+//
+// To use, call `Add()` one or more times to set up the types, facets or
+// interfaces to be iterated over. Then call `Next()` until it returns
+// `Step::Done`.
+class TypeIterator {
+ public:
+  // The result of each iteration step.
+  class Step {
```

Comment:
I kind of hate this variant of variants, and the way it has to be used (a set of overloads). What I want is a tagged union with a discriminant and payloads, and switch statements to consume it. Basically if we had https://suslib.cc/sus-choice_type-Choice.html this available in `//common` I would have used that here. Maybe something to consider, bringing over a copy of something like that?

Reply:
After toolchain discussion we figured we should make CARBON_KIND_SWITCH work with variant and https://github.com/carbon-language/carbon-lang/pull/5433 does that, so these will stay as variants.